### ISSUE TITLE
Problem training on MacOS with cpu device

### ISSUE THREAD

**Issue by antirez** [Om-Alve/smolGPT#20]
Hi again, and sorry for opening two issues in one day! But still I believe this may be interesting.

Apparently when the training starts, with CPU target, I can see many python threads spawning and the macbook fan spinning as expected. Then, after the first step, it writes the checkpoint on disk and stops: it will no longer do any update, the CPU usage drops, nothing more happens, but still a single Python thread uses 100% of CPU. Not sure what this can be. Thanks.

**Comment by antirez** [comment]
P.S. I forgot to say that I'm using caffeine while doing the training, in order to prevent the MacBook to stop. On the same MacBook I was able to end trainings with both Karpathy MiniGPT and other GPT variations I coded myself.

**Comment by antirez** [comment]
<img width="1167" alt="Image" src="https://github.com/user-attachments/assets/ed025cc7-119a-4aed-ab8d-b7ff3a23fd72" />

This is the CPU usage as long as it works correctly. Then when it stops working it drops to 100%.

**Comment by antirez** [comment]
If I train a much smaller model and disable compilation (but didn't try with a small model and compilation enabled) it works, so maybe it's just that it is so slow with the CPU target that, I don't know, maybe it hangs in some of the kernels execution that is incredibly inefficient, probably the backward pass, not sure. So closing... It's just that I need to enable MPS probably :) Thanks.

**Comment by antirez** [comment]
For reference, this is a configuration that you can train in a reasonable amount of time on the CPU.

```
@dataclass
class GPTConfig:
    block_size: int = 128
    vocab_size: int = 4096
    n_layer: int = 4
    n_head: int = 4
    n_embed: int = 128
    dropout: float = 0.2
    bias: bool = False
    use_rotary: bool = False
```

**Comment by Om-Alve** [comment]
Thank you for experimenting with the configurations. I wanted to add support for mps, but I do not own a macbook, so couldn't roll it out. Will do it whenever I get my hands on one!

**Comment by edmundronald** [comment]
I trained from scratch with CPU on Mac  (after setting resume=False) and had exactly the same behavior - the program clearly trains through to one checkpoint  (about 75 mins on my MBP Pro M3) and then the CPU usage goes down to nothing and it just sits there. 

python train.py
Tokens per iteration:  131072
num decayed parameter tensors: 42, with 27,521,024 parameters
num non-decayed parameter tensors: 17, with 8,704 parameters
using fused AdamW: False
Number of params: 27529728
step 0: train_loss 8.4218, val_loss 8.4221
Saving checkpoint to out/
===

Generated Story:

Once upon a timeiqueign avocadoizedels guard want body matter fingerinsinsTT@@ monster shiny crayon rad unlock screen screen every Mamaonely costumeccoli balloons frame frame harsh suc thin laughing queen icy asleepurpectav dig record�� cryst Pete Pete drizz shape later oatthingick pus suit pocket dan danrandpa baseasherasher lampO times sword hidRR rob Mrs haveleaseagesbul hook ways fire theregether pointbin againirstness lake lake tag snack forgetens watching wr fr piece flower birthday fruits means walkshat char sweet stars disgust emplew fear scarfhes Bet gray com butterfly resist bring count bike strawragft flexible dist wallet tidKSorry busyup change change squirrel seat univer sight� be wal emponely land tough pantcletime Mary thread ur mot share~ parade singingctop bandage tooth surprised paws sail thought rockff jumpingous modest modest touc office pre cour env office hook gifted kickedimi smooth Bob Bob from trick resist penguin swingident/ never stri opened ho� C sometimes sometimesfterfter Suddenly view politeimurMcilarryunt clouds�irp tall cabin sit shar ey candy poem penguin fun tightisesaceened soft aeropl magnrandpa stri stK talked couch couch popilled H lightning friend ruinedsting ingredients� bad helpful muguteslf labedient needed kitten beautages

===

**Comment by antirez** [comment]
@edmundronald with the parameters I put above it works:

```
Tokens per iteration:  32768
num decayed parameter tensors: 22, with 1,326,592 parameters                    num non-decayed parameter tensors: 9, with 1,152 parameters
using fused AdamW: False
Number of params: 1327744
step 0: train_loss 8.3515, val_loss 8.3507
Saving checkpoint to out/
iter 0: loss 8.3514, time 120431.26ms
iter 1: loss 8.3483, time 115914.94ms
iter 2: loss 8.3489, time 115584.03ms
iter 3: loss 8.3517, time 116372.33ms

.... after two days ...

step 2470: train_loss 2.1717, val_loss 2.1643
Saving checkpoint to out/
iter 2470: loss 2.3442, time 1133507.49ms
iter 2471: loss 2.3505, time 164694.96ms
iter 2472: loss 2.2547, time 1085995.66ms
iter 2473: loss 2.4112, time 1294275.83ms
iter 2474: loss 2.3837, time 159373.20ms
iter 2475: loss 2.2995, time 1063626.31ms
iter 2476: loss 2.4409, time 1389117.53ms
iter 2477: loss 2.4273, time 159403.57ms
iter 2478: loss 2.3020, time 279096.82ms
```

However I also disabled pytorch compiling in `config.py`, so I'm not sure if this also plays some role with the fact that with a smaller model it works. The model, after the training, works very well (for a model of that size, it's just 1M parameters).

*Once upon a time, there was a little girl named Lily. She loved to play with her toys and eat yummy snacks. One day, she went to the park to play with her friends. While playing, she saw a little boy who was crying. The boy asked Lily if he wanted to help him. Lily said yes and they started to cry. Suddenly, they saw a man with a big bag of candy. The man was very sad and didn't know what to do. Lily thought about what the man said and decided to be a good dog. She decided to keep the candy safe and be careful with it. She was very happy to have a new friends. She learned that she learned that it's important to be kind to others, but she should always remember how she wanted to do things she was, but also a kind. She knew she wanted to the candy and the man was happy. She said, and the man was not scared, but she was still wanted to share the man with her candy.*

I'll retry if the same model works with compilation on.

**Comment by edmundronald** [comment]
@antirez 

Thank you very much for those params, I've just started running them  (MBP3 128G High Power, cpu )  and we'll see what happens. 
The only mod I made to train.py was to set resume=0. The resume logic needs fixing, but sadly I don't program well enough to do it easily. 
Tensorboard only displays a single point so far. 

If it interests you, I'm looking at TinyStories because like MNIST today it's a small convenient experimental platform. One fun thing one could maybe try here is to experiment with quantification, first with the output and llama.c pp :)

So far the cleanest implementation I have found seems to be https://www.reddit.com/r/deeplearning/comments/1d5td9z/einygpt_writing_tiny_stories_with_a_gptesque_llm/

And the fastest https://github.com/pvonderlind/TinyLM


------
iter 100: loss 7.6396, time 75074.76ms
iter 110: loss 7.5268, time 23798.31ms
iter 120: loss 7.4272, time 23979.11ms
iter 130: loss 7.3071, time 24227.99ms
iter 140: loss 7.1824, time 24074.10ms
iter 150: loss 7.0617, time 24172.86ms
iter 160: loss 6.9395, time 24071.19ms
